tests/integration/test_saml.py

```diff
@@ -0,0 +1,102 @@
+#!/usr/bin/env python3
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+"""Discourse integration tests."""
+
+import logging
+import socket
+import unittest.mock
+
+import pytest
+import requests
+import urllib3.exceptions
+from juju.application import Application  # pylint: disable=import-error
+
+logger = logging.getLogger(__name__)
+
+
+@pytest.mark.asyncio
+@pytest.mark.abort_on_fail
+@pytest.mark.usefixtures("setup_saml_config")
+async def test_saml_login(  # pylint: disable=too-many-locals,too-many-arguments
+    app: Application,
+    requests_timeout: int,
+    run_action,
+    setup_saml_config,
+):
+    """
+    arrange: after discourse charm has been deployed, with all required relation established.
+    act: add an admin user and enable force-https mode.
+    assert: user can login discourse using SAML Authentication.
+    """
+    saml_helper = setup_saml_config
+    urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)
+    # discourse need a long password and a valid email
+    # username can't be "discourse" or it will be renamed
+    username = "ubuntu"
+    email = "ubuntu@canonical.com"
+    password = "test-discourse-k8s-password"  # nosecue
+    saml_helper.register_user(username=username, email=email, password=password)
+
+    action_result = await run_action(app.name, "create-user", email=email, admin=True)
```

Comment:
It is interesting that this test is passing. As seen in the file `discourse_rock/patches/discourse-charm.patch`, this means that the rake task `users:exists` is failing someway, otherwise this would not work. 

As a corollary, probably the promote action does not work.

Reply:
found why the test is passing, an error was being caused in some other way. a different return code has been implemented, whcih could have caught this